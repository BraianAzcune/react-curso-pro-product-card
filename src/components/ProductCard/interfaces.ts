import { useProduct } from "../../hooks/useProduct";

export interface Product{
  id: string,
  title: string,
  img?: string;
}

type useProductReturn = ReturnType<typeof useProduct>;
export interface ProductContextProps extends useProductReturn{
  product: Product
}

export interface ProductCardOnChangeArgs {
  product: Product,
  count: number
}