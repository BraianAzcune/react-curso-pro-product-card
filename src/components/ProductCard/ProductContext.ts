import { createContext } from "react";
import {ProductContextProps } from "./interfaces";
export const ProductContext = createContext<ProductContextProps>({} as ProductContextProps);