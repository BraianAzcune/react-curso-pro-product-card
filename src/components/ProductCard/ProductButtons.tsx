import { useContext } from "react";

import styles from "../../styles/styles.module.css";
import React from "react";
import { ProductContext } from "./ProductContext";

export function ProductButtons(props: { className?: string }) {
  const { cantidad, incrementBy, maxCount } = useContext(ProductContext);

  const isMaxReached = maxCount != undefined && cantidad >= maxCount;

  return (
    <div className={`${styles.buttonsContainer} ${props.className}`}>
      <button onClick={() => incrementBy(-1)} className={styles.buttonMinus}>-</button>
      <div className={styles.countLabel}>{cantidad}</div>
      <button
        onClick={() => incrementBy(1)}
        className={`${styles.buttonAdd} ${isMaxReached ? styles.disabled : ""}`}
        disabled={isMaxReached}>+</button>
    </div>
  );
}