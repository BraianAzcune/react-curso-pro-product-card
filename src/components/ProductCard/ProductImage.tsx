import React from "react";
import { useContext } from "react";
import noImage from "../../assets/no-image.webp";
import styles from "../../styles/styles.module.css";

import { ProductContext } from "./ProductContext";

export function ProductImage(props: { img?: string, className?: string }) {
  const { product } = useContext(ProductContext);
  const imagenSrc = props.img ?? product.img ?? noImage;
  return (
    <img className={`${styles.productImg} ${props.className}`} src={imagenSrc} alt="product" />
  );
}