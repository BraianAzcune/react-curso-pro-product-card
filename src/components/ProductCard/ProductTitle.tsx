import { useContext } from "react";

import styles from "../../styles/styles.module.css";
import React from "react";
import { ProductContext } from "./ProductContext";

export function ProductTitle(props: { title?: string, className?: string }) {
  const { product } = useContext(ProductContext);
  const title = props.title ?? product.title;
  return (
    <span className={`${styles.productDescription} ${props.className}`}>{title}</span>
  );
}