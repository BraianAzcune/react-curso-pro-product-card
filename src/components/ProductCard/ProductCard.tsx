import React from "react";

import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";
import { ProductButtons } from "./ProductButtons";
import { ProductContext } from "./ProductContext";
import { InitialValues, useProduct } from "../../hooks/useProduct";
import styles from "../../styles/styles.module.css";
import { Product, ProductCardOnChangeArgs } from "./interfaces";
import { ReactElement } from "react";





export interface ProductCardHandlers {
  count: number,
  isMaxCountReached: boolean,
  maxCount?: number,
  product: Product,
  increaseBy: (v: number) => void;
  reset: () => void,
}

export interface ProductCardProps {
  className?: string,
  product: Product,
  children?: ReactElement | ReactElement[] | ((args: ProductCardHandlers) => JSX.Element),
  style?: React.CSSProperties,
  value?: number,
  onChange?: (args: ProductCardOnChangeArgs) => void
  initialValues?: InitialValues
}

export function ProductCard({ product, children, className, style, value, onChange, initialValues }: ProductCardProps) {

  const { cantidad, incrementBy, maxCount, reset, isMaxCountReached } = useProduct({ value, onChange, product, initialValues });

  return (
    <ProductContext.Provider value={{
      cantidad,
      incrementBy,
      product,
      maxCount,
      reset,
      isMaxCountReached
    }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {/* <ProductImage img={product.img}></ProductImage>
      <ProductTitle title={product.title}></ProductTitle>
      <ProductButtons changeQuantity={incrementBy} quantity={cantidad}></ProductButtons> */}
        {
          typeof children === "function" ?
            children({
              count: cantidad,
              increaseBy: incrementBy,
              isMaxCountReached,
              reset,
              maxCount,
              product,
            })
            :
            children
        }
      </div>
    </ProductContext.Provider>
  );
}



ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Buttons = ProductButtons;

