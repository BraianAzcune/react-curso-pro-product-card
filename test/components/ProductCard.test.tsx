import React from 'react';
import renderer from "react-test-renderer";
import {ProductCard } from "../../src/components/ProductCard";
import { product1 } from './data/products';
const {act} = renderer;

describe("ProductTitle", () => {
    test("debe mostrar el componente", () => {
        const wrapper = renderer.create(
            <ProductCard product={product1}>
                {
                    ()=>(
                        <h1>test producto h1</h1>
                    )
                }
            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot()
    });

    test("debe incrementar el contador", () => {
        const wrapper = renderer.create(
            <ProductCard product={product1}>
                {
                    ({count, increaseBy})=>(
                        <>
                        <h1>test producto card</h1>
                        <span>{count}</span>
                        <button onClick={()=>increaseBy(1)}>incrementar</button>
                        </>
                    )
                }
            </ProductCard>
        )
        let tree = wrapper.toJSON() as any;
        expect(tree).toMatchSnapshot();
        if(!tree.children){
          fail("El test falla porque no se encontraron hijos en el Ã¡rbol renderizado.");
        }
        if(!tree.children[2]){
          fail("El test falla porque no se encontro el tercer hijo que es el boton");
        }
        // las acciones que cambian estados deben estar dentro de act
        act(()=>{
          tree.children[2].props.onClick();
        })
        tree = wrapper.toJSON() as any;
        expect((tree as any).children[1].children[0]).toBe("1")

    });

    // test("debe mostrar el componente con nombre del producto", () => {
    //     const wrapper = renderer.create(
    //         <ProductCard product={product1}>
    //             <ProductCard></ProductCard>
    //         </ProductCard>
    //     )
    //     expect(wrapper.toJSON()).toMatchSnapshot()
    // });

});