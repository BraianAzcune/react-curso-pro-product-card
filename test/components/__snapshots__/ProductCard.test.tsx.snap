// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe incrementar el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    test producto card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    incrementar
  </button>
</div>
`;

exports[`ProductTitle debe mostrar el componente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    test producto h1
  </h1>
</div>
`;
